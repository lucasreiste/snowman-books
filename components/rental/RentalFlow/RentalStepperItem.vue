<template>
  <div
    class="relative flex w-full flex-col items-center justify-center"
    data-testid="stepper-item"
  >
    <div
      v-if="!lastStep"
      class="absolute left-[calc(50%+20px)] right-[calc(-50%+10px)] top-5 block h-0.5 shrink-0 rounded-full bg-muted"
      :class="{
        'bg-primary': state === 'completed',
      }"
      data-testid="stepper-line"
    />
    <Button
      :variant="buttonVariant"
      size="icon"
      class="z-10 rounded-full shrink-0"
      data-testid="stepper-button"
    >
      <LucideCheck
        v-if="state === 'completed'"
        class="size-5"
        data-testid="stepper-icon-completed"
      />
      <LucideCircle
        v-if="state === 'active'"
        data-testid="stepper-icon-active"
      />
      <LucideDot
        v-if="state === 'inactive'"
        data-testid="stepper-icon-inactive"
      />
    </Button>
    <div class="mt-5 flex flex-col items-center text-center">
      <h3
        :class="{ 'text-primary': state === 'active' }"
        class="text-sm font-semibold transition lg:text-base"
        data-testid="stepper-title"
      >
        {{ title }}
      </h3>
      <p
        :class="{ 'text-primary': state === 'active' }"
        class="sr-only text-xs text-muted-foreground transition md:not-sr-only lg:text-sm"
        data-testid="stepper-description"
      >
        {{ description }}
      </p>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { defineProps, computed } from "vue";

const props = defineProps<{
  step: number;
  title: string;
  description: string;
  currentStep: number;
}>();

const state = computed(() => {
  if (props.step < props.currentStep) return "completed";
  if (props.step === props.currentStep) return "active";
  return "inactive";
});

const lastStep = computed(() => {
  return props.step === 4;
});

const buttonVariant = computed(() => {
  if (state.value === "completed" || state.value === "active") {
    return "default";
  }
  return "outline";
});
</script>
