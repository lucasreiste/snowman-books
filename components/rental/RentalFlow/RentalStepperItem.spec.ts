import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/vue";
import RentalStepperItem from "./RentalStepperItem.vue";

const ButtonMock = {
  name: "Button",
  template: `<button :data-testid="$attrs['data-testid']" :variant="variant" :size="size"><slot /></button>`,
  props: ["variant", "size"],
  inheritAttrs: true,
};

describe("RentalStepperItem", () => {
  const renderComponent = (props = {}) => {
    return render(RentalStepperItem, {
      props: {
        step: 1,
        title: "Informações Pessoais",
        description: "Preencha seus dados",
        currentStep: 1,
        ...props,
      },
      global: {
        stubs: {
          Button: ButtonMock,
          LucideCheck: true,
          LucideCircle: true,
          LucideDot: true,
        },
      },
    });
  };

  it("should render correctly", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it("should display title and description correctly", () => {
    renderComponent({
      title: "Teste",
      description: "Descrição teste",
    });

    expect(screen.getByTestId("stepper-title")).toHaveTextContent("Teste");
    expect(screen.getByTestId("stepper-description")).toHaveTextContent(
      "Descrição teste"
    );
  });

  it("should show completed state correctly", () => {
    renderComponent({
      step: 1,
      currentStep: 2,
    });

    expect(screen.getByTestId("stepper-icon-completed")).toBeInTheDocument();
    expect(screen.queryByTestId("stepper-icon-active")).not.toBeInTheDocument();
    expect(
      screen.queryByTestId("stepper-icon-inactive")
    ).not.toBeInTheDocument();
  });

  it("should show active state correctly", () => {
    renderComponent({
      step: 2,
      currentStep: 2,
    });

    expect(
      screen.queryByTestId("stepper-icon-completed")
    ).not.toBeInTheDocument();
    expect(screen.getByTestId("stepper-icon-active")).toBeInTheDocument();
    expect(
      screen.queryByTestId("stepper-icon-inactive")
    ).not.toBeInTheDocument();
  });

  it("should show inactive state correctly", () => {
    renderComponent({
      step: 3,
      currentStep: 2,
    });

    expect(
      screen.queryByTestId("stepper-icon-completed")
    ).not.toBeInTheDocument();
    expect(screen.queryByTestId("stepper-icon-active")).not.toBeInTheDocument();
    expect(screen.getByTestId("stepper-icon-inactive")).toBeInTheDocument();
  });

  it("should not show connector line on last step", () => {
    renderComponent({
      step: 4,
      currentStep: 1,
    });

    const stepperLine = screen.queryByTestId("stepper-line");
    expect(stepperLine).not.toBeInTheDocument();
  });

  it("should show connector line for non-last steps", () => {
    renderComponent({
      step: 2,
      currentStep: 1,
    });

    const stepperLine = screen.getByTestId("stepper-line");
    expect(stepperLine).toBeInTheDocument();
  });

  it("should apply correct button variant for completed state", () => {
    renderComponent({
      step: 1,
      currentStep: 2,
    });

    const button = screen.getByTestId("stepper-button");
    expect(button).toHaveAttribute("variant", "default");
  });

  it("should apply correct button variant for inactive state", () => {
    renderComponent({
      step: 3,
      currentStep: 1,
    });

    const button = screen.getByTestId("stepper-button");
    expect(button).toHaveAttribute("variant", "outline");
  });
});
