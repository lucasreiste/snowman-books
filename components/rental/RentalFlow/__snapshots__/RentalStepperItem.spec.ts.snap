// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RentalStepperItem > should render correctly 1`] = `
<div>
  <div
    class="relative flex w-full flex-col items-center justify-center"
    data-testid="stepper-item"
  >
    <div
      class="absolute left-[calc(50%+20px)] right-[calc(-50%+10px)] top-5 block h-0.5 shrink-0 rounded-full bg-muted"
      data-testid="stepper-line"
    />
    <button
      class="z-10 rounded-full shrink-0"
      data-testid="stepper-button"
      size="icon"
      variant="default"
    >
      
      <!--v-if-->
      <lucide-circle-stub
        data-testid="stepper-icon-active"
      />
      <!--v-if-->
      
    </button>
    <div
      class="mt-5 flex flex-col items-center text-center"
    >
      <h3
        class="text-primary text-sm font-semibold transition lg:text-base"
        data-testid="stepper-title"
      >
        Informações Pessoais
      </h3>
      <p
        class="text-primary sr-only text-xs text-muted-foreground transition md:not-sr-only lg:text-sm"
        data-testid="stepper-description"
      >
        Preencha seus dados
      </p>
    </div>
  </div>
</div>
`;
